client.go

```diff
@@ -728,17 +741,24 @@ func (c *Client) Do(req *Request) (*http.Response, error) {
 		// without racing against the closeBody call in persistConn.writeLoop.
 		httpreq := *req.Request
 		req.Request = &httpreq
+
+		if err := c.PrepareRetry(req.Request); err != nil {
+			prepareErr = err
+			break
+		}
```

Comment:
Per the above, can we conditionally attempt this if the function is non-nil?

```suggestion
		if c.PrepareRetry != nil {
			if err := c.PrepareRetry(req.Request); err != nil {
				prepareErr = err
				break
			}
		}
```